mpcforces_extractor/datastructure/rigids.py

```diff
@@ -93,3 +99,143 @@ def get_subcase_id2part_id2force(self) -> Dict:
+        """
+        if self.axis is not None:
+            return self.axis
+
+        point_cloud = np.array([node.coords for node in self.nodes])
+        centroid = np.array(self.master_node.coords)
+        shifted_points = (
+            point_cloud - centroid
+        )  # Center the points around the master node
+
+        # PCA: Compute the covariance matrix and eigen decomposition
+        covariance_matrix = np.cov(shifted_points, rowvar=False)
+        eigenvalues, eigenvectors = np.linalg.eigh(covariance_matrix)
+
+        # Check if all eigenvalues are too similar (potential for a flat or degenerate structure)
+        if np.isclose(eigenvalues[0], eigenvalues[1], atol=1e-6) and np.isclose(
+            eigenvalues[1], eigenvalues[2], atol=1e-6
+        ):
+            Logger().log_error(
+                f"Element {self.element_id} has nearly identical eigenvalues {eigenvalues}; "
+                f"cylinder axis is ambiguous."
+            )
+            raise ValueError(
+                "Cylinder axis could not be determined; all eigenvalues are too similar."
+            )
+
+        # Identify the outlier eigenvalue
+        mean_of_two_smallest = np.mean(sorted(eigenvalues)[:2])
+        differences = [abs(ev - mean_of_two_smallest) for ev in eigenvalues]
+        outlier_index = np.argmax(differences)
+
+        # Select the eigenvector corresponding to the outlier eigenvalue
+        axis = eigenvectors[:, outlier_index]
+
+        # Handle degenerate cases (e.g., 2D plane of points)
+        if np.isclose(
+            sorted(eigenvalues)[0], 0, atol=1e-6
+        ):  # Detect near-zero variance
+            Logger().log_warn(
+                f"Element {self.element_id} has a 2D plane of nodes; "
+                f"axis will be the cross product of the two dominant eigenvectors"
+            )
+            zero_index = np.argmin(eigenvalues)
+            # cross the other indices
+            axis = np.cross(
+                eigenvectors[:, (zero_index + 1) % 3],
+                eigenvectors[:, (zero_index + 2) % 3],
+            )
+
+        axis /= np.linalg.norm(axis)  # Normalize the axis
+
+        # Save and return the axis
+        self.axis = axis.tolist()
+        self.__compute_diameter_length(shifted_points, axis)
+
+        return self.axis
+
+    def __compute_diameter_length(
+        self, shifted_points: np.array, axis: np.array
+    ) -> None:
+
+        # Project points onto the axis
+        projections = np.dot(shifted_points, axis)
+
+        # Compute length (range of projections along the axis)
+        min_proj = np.min(projections)
+        max_proj = np.max(projections)
+
+        # Compute distances from points to the axis
+        distances = np.linalg.norm(shifted_points - np.outer(projections, axis), axis=1)
+
+        self.length = max_proj - min_proj
+        self.diameter = 2 * np.max(distances)
+
+    def get_part_id2axial_radial_forces(self, subcase: Subcase) -> Dict:
```

Comment:
I still hate this chain mpc - subcase - part - force ... how to resolve ?! Its too much chaining